### ISSUE TITLE
TerminalFormatter with linenos=True ignore "\n" in some token value endswith "\n"

### ISSUE THREAD

**Issue by Anteru** [pygments/pygments-migration-test#854]
(Original issue 1147 created by killkill on 2015-09-10T13:36:22.361734+00:00)

TEST CASE:

```
#!python
from pygments import highlight

_lexer = pygments.lexers.sql.PlPgsqlLexer()

def sql_syntax_highlight(sql):
    highlight(sql,
                     _lexer,
                     pygments.formatters.terminal.TerminalFormatter(
                           linenos=True, bg="dark", encoding="utf-8"))


if __name__ == "__main__":
    sql = """
--    this is Token.Comment.Single
select
N.LAST_APP
from N
    """
    print sql_syntax_highlight(sql.strip())

```

```
#!text

========= OUTPUT BEGIN ===========
0001: --    this is Token.Comment.Single select
0002: N.LAST_APP
0003: from N
========= OUTPUT END   ===========
```

Of course， there is beautiful hightlight in my terminal  :-) 
But , As you see ， the keyword select should display in line 2 , not line 1 

When linenos=False , it works perfectly but no color  :-(

I think something wrong in TerminalFormatter._format_unencoded_with_lineno
I change a little code to work around :

```
#!python

class MyTerminalFormatter(pygments.formatters.terminal.TerminalFormatter):
    def _format_unencoded_with_lineno(self, tokensource, outfile):
        self._write_lineno(outfile)
        for ttype, value in tokensource:

            _need_add_new_line = False
            if value.endswith("\n"):
                self._write_lineno(outfile)
                value = value[:-1]

                # Here is My Changed Code
                if len(value.replace("\n", "")) > 0:
                    _need_add_new_line = True

            color = self.colorscheme.get(ttype)
            while color is None:
                ttype = ttype[:-1]
                color = self.colorscheme.get(ttype)
            if color:
                color = color[self.darkbg]
                spl = value.split('\n')
                for line in spl[:-1]:
                    self._write_lineno(outfile)
                    if line:
                        outfile.write(ansiformat(color, line[:-1]))
                if spl[-1]:
                    outfile.write(ansiformat(color, spl[-1]))
            else:
                outfile.write(value)

            # Here is My Changed Code
            if _need_add_new_line:
                self._write_lineno(outfile)

        outfile.write("\n")
```

**Comment by Anteru** [comment]
(Original issue 1147 last updated on 2015-10-07T13:46:58.483376+00:00)

(Issue automaticaly closed due to status in Bitbucket: resolved)

**Comment by Anteru** [comment]
(Original comment by upgrade_ayp on 2015-10-01T17:45:43.173775+00:00)

I have the same prob. Huang's code works good in my pc win8.1/python 3.4.3

**Comment by Anteru** [comment]
(Original comment by tshatch on 2015-10-07T13:46:55.518955+00:00)

Your script needs some modifications to work, but I confirm the issue.

**Comment by Anteru** [comment]
(Original comment by tshatch on 2015-10-07T13:46:58.445752+00:00)

Address problems with linenos in terminal formatter by rewriting

Fixes #1147

→ <<cset 116d2b88ced4>>
